OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/BaseBuilderQueueManager.cs

```diff
@@ -287,62 +298,61 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
-				if (count * 100 > frac.Value * playerBuildings.Length)
-					continue;
-
-				if (baseBuilder.Info.BuildingLimits.ContainsKey(name) && baseBuilder.Info.BuildingLimits[name] <= count)
-					continue;
-
-				// If we're considering to build a naval structure, check whether there is enough water inside the base perimeter
-				// and any structure providing buildable area close enough to that water.
-				// TODO: Extend this check to cover any naval structure, not just production.
-				if (baseBuilder.Info.NavalProductionTypes.Contains(name)
-					&& (waterState == WaterCheck.NotEnoughWater
-						|| !AIUtils.IsAreaAvailable<GivesBuildableArea>(world, player, world.Map, baseBuilder.Info.CheckForWaterRadius, baseBuilder.Info.WaterTerrainTypes)))
-					continue;
-
-				// Will this put us into low power?
-				var actor = world.Map.Rules.Actors[name];
-				if (playerPower != null && (playerPower.ExcessPower < minimumExcessPower || !HasSufficientPowerForActor(actor)))
+				int owned = 0;
+				ownedActorCounts.TryGetValue(b.Name, out owned);
+
+				var desiredPercentage = queueConsiderations[b.Name].Percentage;
+				if (owned * 100 < desiredPercentage * queueTotal)
 				{
-					// Try building a power plant instead
-					if (power != null && power.TraitInfos<PowerInfo>().Where(i => i.EnabledByDefault).Sum(pi => pi.Amount) > 0)
-					{
-						if (playerPower.PowerOutageRemainingTicks > 0)
-							AIUtils.BotDebug("{0} decided to build {1}: Priority override (is low power)", queue.Actor.Owner, power.Name);
-						else
-							AIUtils.BotDebug("{0} decided to build {1}: Priority override (would be low power)", queue.Actor.Owner, power.Name);
+					// If we're considering to build a naval structure, check whether there is enough water inside the base perimeter
+					// and any structure providing buildable area close enough to that water.
+					// TODO: Extend this check to cover any naval structure, not just production.
+					if (baseBuilder.Info.NavalProductionTypes.Contains(b.Name)
+					    && (waterState == WaterCheck.NotEnoughWater
+					        || !AIUtils.IsAreaAvailable<GivesBuildableArea>(world, player, world.Map, baseBuilder.Info.CheckForWaterRadius, baseBuilder.Info.WaterTerrainTypes)))
+						continue;
 
-						return power;
+					// Will this put us into low power?
+					var actor = world.Map.Rules.Actors[b.Name];
+					if (playerPower != null && (playerPower.ExcessPower < minimumExcessPower || !HasSufficientPowerForActor(actor)))
+					{
+						// Try building a power plant instead
+						if (power != null && power.TraitInfos<PowerInfo>().Where(i => i.EnabledByDefault).Sum(pi => pi.Amount) > 0)
+						{
+							if (playerPower.PowerOutageRemainingTicks > 0)
+								AIUtils.BotDebug("{0} decided to build {1}: Priority override (is low power)", queue.Actor.Owner, power.Name);
+							else
+								AIUtils.BotDebug("{0} decided to build {1}: Priority override (avoid low power)", queue.Actor.Owner, power.Name);
```

Comment:
The above comment about false debug messages applies here too, surprisingly when explicitly telling the AI to build a power plant. Since I'm using the default value for `minimumExcessPower` of `0`, it looks like something is not right with `HasSufficientPowerForActor` when building power plants.